From f855623541b27ee83f4c9542fea484db62da1df5 Mon Sep 17 00:00:00 2001
From: Andy Ross <andy.ross@windriver.com>
Date: Tue, 28 May 2013 11:02:25 -0700
Subject: Allow for overriding environment variables in export

Don't silently append duplicates, delete the older version.  Also
increase the size limit, as the set of shell variables on my test
system is already at 28.

Change-Id: Id2373cc197662ae14fabad4de5e4516fe30d44cc
For: AXIA-2870
Signed-off-by: Andy Ross <andy.ross@windriver.com>
---
 init/init.c |   18 ++++++++++++++----
 1 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/init/init.c b/init/init.c
index 43b2b9e..0b5fe1a 100755
--- a/init/init.c
+++ b/init/init.c
@@ -93,14 +93,24 @@ static int have_console;
 static char *console_name = "/dev/console";
 static time_t process_needs_restart;
 
-static const char *ENV[32];
+static const char *ENV[128];
 
 /* add_environment - add "key=value" to the current environment */
 int add_environment(const char *key, const char *val)
 {
-    int n;
-
-    for (n = 0; n < 31; n++) {
+    unsigned int n, kesz;
+    char keyeq[128];
+    int max_vars = sizeof(ENV)/sizeof(ENV[0]) - 1; /* final null for execve */
+
+    snprintf(keyeq, sizeof(keyeq), "%s=", key);
+    kesz = strlen(keyeq);
+
+    for (n = 0; n < max_vars; n++) {
+        if (ENV[n] && strncmp(ENV[n], keyeq, kesz) == 0) {
+            /* Override */
+            free((char*)ENV[n]);
+            ENV[n] = NULL;
+        }
         if (!ENV[n]) {
             size_t len = strlen(key) + strlen(val) + 2;
             char *entry = malloc(len);
-- 
1.7.4.1

